### PR TITLE
[tests] Allow running only selected apk tests

### PR DESCRIPTION
**Pull request by radekdoulik**
Make it possible to run selected apk tests by passing `TESTS` variable
to `make`.

Example usage:

    make run-apk-tests TESTS="Xamarin.Forms_Performance_Integration;Xamarin.Android.Locale_Tests"

Or msbuild equivalent:
    
        msbuild /t:RunApkTests tests/RunApkTests.targets /p:ApkTests='"Xamarin.Forms_Performance_Integration;Xamarin.Android.Locale_Tests"'
    
The relevant part of the output of the msbuild example above looks like:
    
        UndeployTestApks:
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   uninstall "Xamarin.Android.Locale_Tests"
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   uninstall "Xamarin.Forms_Performance_Integration"
        DeployTestApks:
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   install "/Users/rodo/git/xa5/tests/../bin/TestDebug/Xamarin.Android.Locale_Tests-Signed.apk"
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   install "/Users/rodo/git/xa5/tests/../bin/TestDebug/Xamarin.Forms_Performance_Integration-Signed.apk"
        RecordApkSizes:
          stat -f "stat: %z %N" "/Users/rodo/git/xa5/tests/../bin/TestDebug/Xamarin.Forms_Performance_Integration-Signed.apk" > "/Users/rodo/git/xa5/tests/../bin/TestDebug/apk-sizes-Xamarin.Forms_Performance_Integration-Debug.txt"
          unzip -l "/Users/rodo/git/xa5/tests/../bin/TestDebug/Xamarin.Forms_Performance_Integration-Signed.apk" >> "/Users/rodo/git/xa5/tests/../bin/TestDebug/apk-sizes-Xamarin.Forms_Performance_Integration-Debug.txt"
        RunTestApks:
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   shell pm grant Xamarin.Android.Locale_Tests android.permission.READ_EXTERNAL_STORAGE
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   shell pm grant Xamarin.Android.Locale_Tests android.permission.WRITE_EXTERNAL_STORAGE
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   shell am instrument  -e "loglevel Verbose" -w "Xamarin.Android.Locale_Tests/xamarin.android.localetests.TestInstrumentation"
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   logcat -v threadtime -d
            appending stdout to file: logcat-Debug-Xamarin.Android.Locale_Tests.txt
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   logcat -c
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   pull "/storage/emulated/0/Android/data/Xamarin.Android.Locale_Tests/files/Documents/TestResults.xml" "/Users/rodo/git/xa5/tests/../bin/TestDebug/TestResult-Xamarin.Android.Locale_Tests.xml"
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   shell am start -n "Xamarin.Forms_Performance_Integration/xamarin.forms.performance.integration.MainActivity"
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   logcat -v threadtime -d
            appending stdout to file: logcat-Debug-Xamarin.Forms_Performance_Integration.txt
          Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb   logcat -c

### CODE REVIEW

**Comment by jonpryor** [comment]
Overall: how often do you want to run *more than one* `.apk` but not all of them?

We can already [run a single `.apk`](https://github.com/xamarin/xamarin-android/blob/master/Documentation/building/unix-instructions.md#running-individual-apk-projects).  It's slightly more verbose, but i don't think it's terribly so?

File: tests/RunApkTests.targets [hunk 1]
```diff
@@ -35,6 +43,28 @@
       ReportComponentFailures;
     </RunApkTestsDependsOn>
   </PropertyGroup>
+
+  <Target Name="FilterApkTests">
+    <ItemGroup Condition=" '$(ApkTests)' == '' ">
+      <TestApkToRun Include="@(TestApk)" />
+    </ItemGroup>
+    <ItemGroup Condition=" '$(ApkTests)' == '' ">
```
Comment (jonpryor) [inline tests/RunApkTests.targets]: This (and the following item groups) are what really give me pause.
```suggestion
    <ItemGroup Condition=" '$(ApkTests)' == '' ">
```
```suggestion
    <ItemGroup Condition=" '$(ApkTests)' == '' ">
```

Consider: https://github.com/xamarin/xamarin-android/blob/13a692511ad77c1494c2e9aaff84d7d60efb0de0/src/Mono.Android/Test/Mono.Android-Tests.projitems#L20-L33

`Mono.Android-Tests.csproj` sets `@(TestApkInstrumentation)` and `@(TestApkPermission)`.

With this PR, if you run `make run-apk-tests TESTS="Mono.Android_Tests"`, then *neither* of these item groups are evaluated, meaning permissions *won't* be set (afaict), and `<RunInstrumentationTests/>` won't be executed either (afaict; if it *would* be executed, I don't currently see how/why).

Without instrumentation, this makes for a very "weird" interpretation of `run-apk-tests`, when we'd just be *launching* the test runners, not *executing the tests*.

Comment (radekdoulik) [reply tests/RunApkTests.targets]: I am not sure why they wouldn't. Here's the relevant part of the build output of `make run-apk-tests V=1 CONFIGURATION=Release TESTS="Mono.Android_Tests"`:
```
DeployTestApks:
  Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb -s emulator-5570  install "/Users/rodo/git/xa5/tests/../bin/TestRelease/Mono.Android_Tests-Signed.apk"
RecordApkSizes:
  stat -f "stat: %z %N" "/Users/rodo/git/xa5/tests/../bin/TestRelease/Mono.Android_Tests-Signed.apk" > "/Users/rodo/git/xa5/tests/../bin/TestRelease/apk-sizes-Mono.Android_Tests-Release.txt"
  unzip -l "/Users/rodo/git/xa5/tests/../bin/TestRelease/Mono.Android_Tests-Signed.apk" >> "/Users/rodo/git/xa5/tests/../bin/TestRelease/apk-sizes-Mono.Android_Tests-Release.txt"
RunTestApks:
  Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb -s emulator-5570  shell pm grant Mono.Android_Tests android.permission.READ_EXTERNAL_STORAGE
  Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb -s emulator-5570  shell pm grant Mono.Android_Tests android.permission.WRITE_EXTERNAL_STORAGE
  Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb -s emulator-5570  shell am instrument  -e "loglevel Verbose" -w "Mono.Android_Tests/xamarin.android.runtimetests.TestInstrumentation"
  Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb -s emulator-5570  logcat -v threadtime -d
    appending stdout to file: logcat-Release-Mono.Android_Tests.txt
  Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb -s emulator-5570  logcat -c
  Executing: /Users/rodo/android-toolchain/sdk/platform-tools/adb -s emulator-5570  pull "/storage/emulated/0/Android/data/Mono.Android_Tests/files/Documents/TestResults.xml" "/Users/rodo/git/xa5/tests/../bin/TestRelease/TestResult-Mono.Android_Tests.xml"
```

This case it handled on lines 57-65 bellow.

Comment (radekdoulik) [reply tests/RunApkTests.targets]: > Overall: how often do you want to run more than one .apk but not all of them?

My motivation here is to be able to run faster the selected test(s) with the performance and size measurements. I found myself too often in that situation, so I got annoyed by it and wrote that PR :-)

Comment (radekdoulik) [reply tests/RunApkTests.targets]: BTW, I have an idea how to make it more elegant by using conditional item group removal.

If that works, we can avoid creating the new `*ToRun` groups, so the change would became a lot simpler.

Comment (radekdoulik) [reply tests/RunApkTests.targets]: OK, updated. The removal version looks much nicer IMO.

**Comment by jonpryor** [comment]
Let's use a `PACKAGES` variable instead of a `TESTS` variable, and please add this new mechanism to the documentation.

**Pull request merged**